src/switchboard.rs

```diff
@@ -255,17 +263,24 @@ impl Switchboard {
     }
 
     pub fn get_publisher(&self, user_id: &UserId) -> Option<&Arc<Session>> {
-        self.sessions
-            .iter()
-            .find(|s| {
-                let subscriber_offer = s.subscriber_offer.lock().unwrap();
-                let join_state = s.join_state.get();
-                match (subscriber_offer.as_ref(), join_state) {
-                    (Some(_), Some(state)) if &state.user_id == user_id => true,
-                    _ => false,
-                }
-            })
-            .map(Box::as_ref)
+        let mut result = None;
+
+        if let Some(room_id) = self.users_to_room.get(user_id) {
+            if let Some(sessions) = self.occupants.get(room_id) {
+                result = sessions
+                    .iter()
+                    .find(|s| {
+                        let subscriber_offer = s.subscriber_offer.lock().unwrap();
+                        let join_state = s.join_state.get();
+                        match (subscriber_offer.as_ref(), join_state) {
+                            (Some(_), Some(state)) if &state.user_id == user_id => true,
+                            _ => false,
+                        }
+                    })
```

Comment:
Note that now we're not returning a reference to the boxes session in `self.sessions` - I'm assuming this is OK since we're still returning a reference (ie we won't be copying the sessions on each call, right?)

Reply:
This should be fine in this context. There's actually some nuance here that you don't understand because it's confusing and I didn't document it very well. I might as well explain it here because it's important that I explain it somewhere.

When Janus calls into our plugin, and when our plugin calls back to Janus, Janus always passes and accepts a pointer to a [`janus_plugin_session`](https://github.com/meetecho/janus-gateway/blob/master/plugins/plugin.h#L237) struct ([`PluginSession`](https://github.com/mozilla/janus-plugin-rs/blob/master/janus-plugin-sys/src/plugin.rs#L35) [type](https://github.com/mozilla/janus-plugin-rs/blob/master/src/lib.rs#L13) in Rust FFI.) This struct has a field `plugin_handle` which is a pointer that is free for our plugin to use. The intent is that any state associated with a session should well be put here, instead of having it in a separate data structure that you need to look up all the time.

A natural thing to do would be to make a kind of `SessionState` struct and then set the `plugin_handle` to be a pointer to a `SessionState`. This has one hitch, which is that you very often in the plugin code need to get back the `janus_plugin_session` corresponding to the state, to refer to that session to Janus; for example, if you want to send an RTP packet to that session. So the `SessionState` also needs to point back to the `janus_plugin_session` it belongs to.

There's an additional concern, which is that since plugins are going to need to store the `janus_plugin_session`s they are given, the plugins are made responsible to participate in their memory lifecycle through Janus's reference counts. That is, the plugin needs to call `decrease` on the `janus_plugin_session`'s refcount when it's done, which will free it once Janus core is also done with it.

In Rust I wrote the `[SessionWrapper<T>](https://github.com/mozilla/janus-plugin-rs/blob/master/src/session.rs#L25)` type to encapsulate this. A `SessionWrapper<T>` is a (`handle`, T) pair made to be a wrapper around a reference-counted `janus_plugin_session` pointer (`handle`) whose `plugin_handle` stores state of type T. It implements `Deref<T>` so that you can treat it like a T, but you can also get the `handle` from it, and when it drops in Rust, the reference count is decreased. In our plugin, the `Session` type is defined as a [`SessionWrapper<SessionState>`.](https://github.com/mozilla/janus-plugin-sfu/blob/master/src/sessions.rs#L44) The [`SessionState`](https://github.com/mozilla/janus-plugin-sfu/blob/master/src/sessions.rs#L25) type is what has fields like `join_state`, etc.

To use it, you call [`SessionWrapper.associate(handle, state)`](https://github.com/mozilla/janus-plugin-rs/blob/master/src/session.rs#L48) on a `*janus_plugin_session` and a `T`, and you get back a `Box<Arc<SessionWrapper<T>>>`. Each of these wrappers is meaningful. The `Arc` is to make the session wrapper be Rust-reference-counted, so that it's easy to stick it in the switchboard, pass it around threads, and so on, and the `SessionWrapper` will go away only when you're really done with it. The `Box` is allocated so that the inner `plugin_handle` has something on the heap to point to -- the `plugin_handle` field points to the box. This box is what lives in the switchboard `sessions` vector -- it's the raw storage for all of these. So the `sessions` vector is a little special and that's why it has the weird looking box.

In this case, we're just looking up the publisher and doing some stuff based on the state, so there's no reason why returning the box is important. Returning the `Arc` inside the box is best.

Sorry if this seems confusing, it's definitely the most complicated part of the plugin!